File: FrontEnd/lib/data/mappers/message_mapper.dart
/// @file message_mapper.dart
/// @description Implementación del mapper que transforma MessageModel ↔ Message.
/// @module Chat
/// @layer Data
library;

import '../../domain/entities/message.dart';
import '../i_mappers/i_message_mapper.dart';
import '../models/message_model.dart';

// ═══════════════════════════════════════════════════════════════════════════════
// MESSAGE MAPPER
// ═══════════════════════════════════════════════════════════════════════════════

/// Convierte entre [MessageModel] (serialización JSON) y [Message] (entidad de dominio).
///
/// [implements] IMessageMapper
/// [injectable] registrar en container.dart como IMessageMapper.
class MessageMapper implements IMessageMapper {
  const MessageMapper();

  /// Convierte un [MessageModel] deserializado del JSON de backend en [Message].
  ///
  /// El campo status se mapea del String ('pending'|'delivered'|'read') al enum [MessageStatus].
  ///
  /// [param] model — modelo de serialización del mensaje.
  /// [returns] entidad de dominio limpia.
  @override
  Message toEntity(MessageModel model) {
    return Message(
      id:             model.id,
      conversationId: model.conversationId,
      senderId:       model.senderId,
      senderName:     model.senderName,
      text:           model.text,
      status:         _parseStatus(model.status),
      tempId:         model.tempId,
      createdAt:      DateTime.parse(model.createdAt).toUtc(),
    );
  }

  /// Convierte una [Message] de dominio en [MessageModel] (para caché o cola offline).
  ///
  /// [param] entity — entidad de dominio del mensaje.
  /// [returns] modelo de serialización.
  @override
  MessageModel toModel(Message entity) {
    return MessageModel(
      id:             entity.id,
      conversationId: entity.conversationId,
      senderId:       entity.senderId,
      senderName:     entity.senderName,
      text:           entity.text,
      status:         entity.status.name,
      tempId:         entity.tempId,
      createdAt:      entity.createdAt.toIso8601String(),
    );
  }

  // ─── Helpers ──────────────────────────────────────────────────────────────

  /// Convierte un String de estado en el enum [MessageStatus].
  ///
  /// [param] raw — String del backend ('pending', 'sent', 'delivered', 'read').
  /// [returns] enum correspondiente; 'sent' por defecto si es desconocido.
  MessageStatus _parseStatus(String raw) {
    switch (raw) {
      case 'pending':
        return MessageStatus.pending;
      case 'sent':
        return MessageStatus.sent;
      case 'delivered':
        return MessageStatus.delivered;
      case 'read':
        return MessageStatus.read;
      default:
        return MessageStatus.sent;
    }
  }
}

File: BackEnd/src/data/mappers/message_mapper.ts
/**
 * @file message_mapper.ts
 * @description Implementación del mapper de mensajes de chat.
 * Convierte entre MessageDocument (MongoDB), Message (dominio) y MessageResponseDTO (presentación).
 * @module Chat
 * @layer Data
 *
 * @implements {IMessageMapper}
 * @injectable
 */

import { injectable } from 'inversify';
import { ObjectId } from 'mongodb';
import type { IMessageMapper } from '../IMappers/IMessageMapper.js';
import { Message, ContentMeta } from '../../domain/entities/Message.js';
import type { MessageDocument } from '../datasources/mongodb/models/MessageModel.js';
import type { MessageResponseDTO } from '../../domain/dtos/chat/message-response.dto.js';

/**
 * Mapper de mensajes de chat.
 *
 * @implements {IMessageMapper}
 * @injectable
 */
@injectable()
export class MessageMapper implements IMessageMapper {

  /**
   * Convierte un documento MongoDB a entidad Message.
   *
   * @param doc — Documento de la colección 'messages'.
   * @returns Entidad Message.
   */
  toEntity(doc: MessageDocument): Message {
    return new Message({
      id:             doc._id.toHexString(),
      conversationId: doc.conversationId.toHexString(),
      senderId:       doc.senderId.toHexString(),
      receiverId:     doc.receiverId?.toHexString(),
      text:           doc.text,
      contentMeta:    doc.contentMeta,
      status:         doc.status,
      tempId:         doc.tempId,
      createdAt:      doc.createdAt,
      updatedAt:      doc.updatedAt,
    });
  }

  /**
   * Construye un documento MongoDB para insertar un nuevo mensaje (sin _id).
   * El estado inicial es siempre 'sent' (el servidor ya ha recibido el mensaje).
   *
   * @param conversationId — ID de la conversación.
   * @param senderId — ID del emisor.
   * @param receiverId — ID del receptor.
   * @param text — Texto del mensaje.
   * @param contentMeta — Metadatos de contenido multimedia.
   * @param tempId — ID temporal del cliente.
   * @returns Documento parcial para insertar en MongoDB.
   */
  toDocument(
    conversationId: string,
    senderId: string,
    receiverId: string,
    text?: string,
    contentMeta?: ContentMeta,
    tempId?: string,
  ): Omit<MessageDocument, '_id'> {
    const now = new Date();
    return {
      conversationId: new ObjectId(conversationId),
      senderId:       new ObjectId(senderId),
      receiverId:     receiverId ? new ObjectId(receiverId) : undefined,
      text,
      contentMeta,
      status:         'sent',
      tempId,
      createdAt:      now,
      updatedAt:      now,
    };
  }

  /**
   * Convierte una entidad Message al DTO de respuesta HTTP, enriquecido con el nombre del emisor.
   *
   * @param entity — Entidad Message.
   * @param senderName — Nombre del emisor.
   * @returns MessageResponseDTO serializable.
   */
  toResponseDTO(entity: Message, senderName: string): MessageResponseDTO {
    return {
      id:             entity.id,
      conversationId: entity.conversationId,
      senderId:       entity.senderId,
      senderName,
      text:           entity.text,
      contentMeta:    entity.contentMeta,
      status:         entity.status,
      tempId:         entity.tempId,
      createdAt:      entity.createdAt.toISOString(),
    };
  }
}
